src/Reflection/ParameterReflection.php

```diff
@@ -58,34 +60,39 @@ public function getClass()
     #[ReturnTypeWillChange]
     public function getDeclaringFunction()
     {
-        $phpReflection = parent::getDeclaringFunction();
-        if ($phpReflection instanceof ReflectionMethod) {
-            $laminasReflection = new MethodReflection($this->getDeclaringClass()->getName(), $phpReflection->getName());
-        } else {
-            $laminasReflection = new FunctionReflection($phpReflection->getName());
+        $function = parent::getDeclaringFunction();
+
+        if ($function instanceof ReflectionMethod) {
+            return new MethodReflection($function->getDeclaringClass()->getName(), $function->getName());
         }
-        unset($phpReflection);
 
-        return $laminasReflection;
+        return new FunctionReflection($function->getName());
     }
 
     /**
      * Get parameter type
      *
+     * @deprecated this method is unreliable, and should not be used: it will be removed in the next major release.
+     *             It may crash on parameters with union types, and will return relative types, instead of
+     *             FQN references
+     *
      * @return string|null
      */
     public function detectType()
     {
         if (
-            method_exists($this, 'getType')
-            && null !== ($type = $this->getType())
+            null !== ($type = $this->getType())
```

Comment:
Assignment in a conditional always makes me uncomfortable. Could this be:

```suggestion
        $type = $this->getType();
        if (null !== $type && $type->isBuiltin()) {
```

Reply:
I didn't want to work on `detectType` any further: it was deprecated in this patch because after throwing more than 30 minutes at it, I raged :D